### ISSUE TITLE
Using wavelets in FNet instead of FFT

### ISSUE THREAD

**Issue by snakers4** [v0lta/PyTorch-Wavelet-Toolbox#10]
Hi @v0lta ,

My name is Alexander, I am one of the authors of https://github.com/snakers4/silero-models.

Some time ago there was a [paper](http://arxiv.org/abs/2105.03824) by Google (FNet: Mixing Tokens with Fourier Transforms) , which essentially took a self-attention transformer module and replaced the self-attention mechanism with an FFT layer and it allegedly reduced computation a lot without affection metrics a lot.

On the other hand, in PyTorch it is as simple as just using this [method](https://pytorch.org/docs/1.9.0/generated/torch.fft.fft2.html) instead of the actual self-attention layer. The data is essentially `batch_size * sequence_length * hidden_dimension`.

I tried this idea in my domain (speech to text) and it really worked to my astonishment (!). Of course I applied it to a small network (aiming to make it much faster, remove SoftMax layer and compress the network 2-3x both in terms of speed and parameters). I could not get anywhere near the boasted by Google mere 10% quality reduction, but probably and surprisingly, ceteris paribus, the 2x smaller network (and probably much faster due to the most expensive part being removed) could boast maybe 25-30% worse metrics.

I had and idea, maybe try wavelet transform instead of FFT? Since they are supposed to be localized both in time and frequency. Then I googled a bit, found `pywt` and then your library. I am not very well acquainted with wavelets, so I found a `wavedec2` function on your README and I would like to ask a couple of questions:

- If I run `wavelist()` I get a long list of filters. Since I am not very familiar with them, could you maybe point me somewhere where I can find some intuition behind choosing them?;

- Do gradients flow through your functions, i.e. can they be used as feature extractors in the middle of the network? In your learnable example you manually construct the network, but you learn the filters themselves, not use them as layers;

**Comment by v0lta** [comment]
Dear @snakers4 ,
- Choosing the right wavelet is a little bit of an art. I recommend a Daubechies (```db```, in ```wavelist()```) Wavelet or a Symlet (```sym```), which is a variant of the Daubechies wavelet family with added symmetry. Both families usually work well in most cases. Choosing the degree, that is the number behind the family abbreviation, depends on the input size and the scale you want.
The filter length of your Wavelet will be the degree times two, i.e. ```len(pywt.Wavelet('db2'))``` will return 4. If you work with ```wavedec2``` the input signal is padded on the edges, to ensure every wavelet-filter-value touches every pixel. Since the stride of the fwt convolutions is two, Haar-Wavelets (```haar``` or ```db1```) require no padding. Longer wavelets can extract more complex relationships but require more padding. The padding, in turn, limits the number of scales that can be computed.  Further decompositions don't make sense, when all that's left is padding. Choosing the right wavelet and degree requires a little bit of trial and error. I usually start with a Haar Wavelet and optimize this later, when I have a working prototype.
- Gradients do flow through all functions. The gradients are required to make the wavelet-filter optimization idea in https://github.com/v0lta/PyTorch-Wavelet-Toolbox/blob/main/examples/network_compression/mnist_compression.py work. You dont have to use the gradients to optimize the wavelet filter tough. If you pass a ```pywt.Wavelet``` to ```wavedec2``` the filters will stay constant, yet gradients still flow through the function for later use.
- I like your idea, I did something similar by replacing the Welsh-Hadamard transform, a real FFT variant, from https://openaccess.thecvf.com/content_iccv_2015/papers/Yang_Deep_Fried_Convnets_ICCV_2015_paper.pdf  with a fast wavelet-transform in https://arxiv.org/pdf/2004.09569.pdf , which worked well. In your https://arxiv.org/pdf/2105.03824.pdf case, the wavelet transform may be particularly useful since neighbouring letters or words will be processed together. You can choose the granularity by setting the scale or level of your fwt, which may be an interesting subject to investigate.

**Comment by snakers4** [comment]
Many thanks for your advice.
Since you say that gradients flow through your filters, I will definitely give it it a try in near future.

From what you say and the docs it seems that I cannot use `wavedec`, so I need to use `wavedec2`, which looks like treats the data like a grayscale image (or I need to use wavelets as the first feature extractor instead of STFT on audio directly, but that would imply lengthy retraining from scratch which is a pain as usual):

```
wavedec2 data (torch.tensor): [batch_size, 1, height, width]
wavedec data (torch.tensor): Input time series of shape [batch_size, 1, time]
```

Am I doing it right?
My typical feature map inside of the network is typically for example `10 * 30 * 512` (batch * sequence * hidden dimension). And when I apply self-attention and / or FFT2 the dimensions are just kept the same.

```
import pywt
import ptwt
import torch
import numpy as np

data_torch = torch.randn(10, 1, 30, 512) 
# a typical feature map size for my use case
# though I have 1D data
print(data_torch.size())
wavelet = pywt.Wavelet('db1')

out = ptwt.wavedec2(data_torch,
                    wavelet,
                    level=2,
                    mode='reflect')

for i, _ in enumerate(out):
    if type(_) == tuple:
        for j, __ in enumerate(_):
            print(f'Level {i},{j} {__.size()}')
    else:
        print(f'Level {i} {_.size()}')
```
when I apply these fitlers, I get:

```
Level 0 torch.Size([10, 1, 8, 128])
Level 1,0 torch.Size([10, 1, 8, 128])
Level 1,1 torch.Size([10, 1, 8, 128])
Level 1,2 torch.Size([10, 1, 8, 128])
Level 2,0 torch.Size([10, 1, 15, 256])
Level 2,1 torch.Size([10, 1, 15, 256])
Level 2,2 torch.Size([10, 1, 15, 256])
```
I understand why I get several layers of outputs, but why the first and second have the same length (128) and why second and third have 3 feature maps inside? I am a bit confused why I get so many feature maps instead of 1 or 2.

**Comment by v0lta** [comment]
You are doing it right. Your level numbers are inverted tough. It's the pywt convention. I am sorry, the documentation is still a bit sparse. So you cant tell by just looking at my docstrings. 
Please take a look at https://pywavelets.readthedocs.io/en/latest/ref/2d-dwt-and-idwt.html , the ptwt.wavedec2 code does the same. But padding options are limited to what torch.nn.functional.pad supports , pywt offers a little bit more here. The decomposition axes are fixed, you got that right though. 

If you flip your level numbering convention the output you got makes sense. Level one has half the resolution of the original input, and level two has half of level 1. 15 was an odd number so a zero was padded to make it 16, divided by two makes that 8.

**Comment by snakers4** [comment]
> If you flip your level numbering convention the output you got makes sense. Level one has half the resolution of the original input, and level two has half of level 1. The height of 15 was an odd number, so a zero (or reflection) was padded to make it 16, divided by two makes that 8.
> Please take a look at https://pywavelets.readthedocs.io/en/latest/ref/2d-dwt-and-idwt.html , the ptwt.wavedec2 code does the same.

I see, many thanks.
Also these pages make everything much more clear - https://pywavelets.readthedocs.io/en/latest/index.html, https://pywavelets.readthedocs.io/en/latest/ref/2d-decompositions-overview.html, maybe it is worth linking them somewhere.

Now I see why there are 3 feature maps - because essentially 1D filters are applied to 2D images.

> The decomposition axes are fixed

`data_torch = torch.randn(10, 1, 30, 512) ` 

there is no getting around this `1` because you are using Conv2D under the hood right?

> ![image](https://user-images.githubusercontent.com/12515440/135395946-77eef2fb-1e09-4390-851c-db7da6ea1bb4.png)
> ![image](https://user-images.githubusercontent.com/12515440/135396104-1e82504b-6d6e-448f-9118-21f6a5bdf774.png)

I wonder how I can incorporate the multi-level features into the pre-existing schema, i.e. make the shape of input equal the shape of output. I wonder if I can just use `coeffs_to_array` from `ptwt` (but that would break the "meaning" of the axes of feature-maps).

But I also need to preserve the meaning of axes, i.e. "time" and "channels".  Maybe I can splice together the horizontal and diagonal feature maps and then apply and expansion layer (256 back to 512). I need to think about it.

Ideally to best use the strength of this type of features, I need to use at least 2 levels, but I cannot think of an easy scheme to do so.

**Comment by v0lta** [comment]
> there is no getting around this 1 because you are using Conv2D under the hood right?
Exactly. Conv2D is also the reason the axes are fixed. However, this should not be too big of a limitation, since transposing the axes before calling ```ptwt.wavedec2``` should always be possible.

Please tell me more about the spectrogram you shared. Was this computed using a short-time Fourier transform? The wavelet analogue to the STFT is the wavelet-packet-transform, I have an example here: https://github.com/v0lta/PyTorch-Wavelet-Toolbox/tree/main/examples/wavelet_packet_chirp_analysis
If you want equal shapes you definitely want a packet transform. I have a 2d forward version here: https://github.com/v0lta/PyTorch-Wavelet-Toolbox/blob/fa7c6ba97b1b3c6197ddd94006435a815eb22df1/src/ptwt/packets.py#L58

**Comment by v0lta** [comment]
So the plan is to convolve in time _and_ frequency, right? In that case your graycode is too short. For the ```natural``` order you can look through all filter combinations:
```
    import pywt
    import torch
    from ptwt import WaveletPacket2D
    from itertools import product
     
    wavelet = pywt.Wavelet('db2')
    x = torch.randn(10, 30, 512)
    wp = WaveletPacket2D(data=x,
                         wavelet=wavelet,
                         mode='reflect')
     
    wp_keys = list(product(["a", "h", "v", "d"], repeat=1))
    np_lst = []
    for key in wp_keys:
         np_lst.append(wp["".join(key)])
    viz = torch.stack(np_lst)
    print(viz.shape)
``` 
if prints ```torch.Size([4, 10, 1, 16, 257])```, which is expected, because the 2d-case uses four filters more here:
https://pywavelets.readthedocs.io/en/latest/regression/wp2d.html#traversing-wp-tree
The gray code is just another way to order the packets for Human consumption. if you are planning to feed the data into a network layer it is not important. I don't think you need the graycode, but if you must pywt implements it correctly here: https://github.com/PyWavelets/pywt/blob/1fb6835132b84b507e7459b2ec15ee5d1c92a1b8/pywt/_wavelet_packets.py#L687

**Comment by v0lta** [comment]
To keep the shape the same as Fourier would, concatenate the packets and use a padding free wavelet like ```db1```, for example:
```
     import pywt
     import torch
     from ptwt import WaveletPacket2D
     from itertools import product
 
     wavelet = pywt.Wavelet('db1')
     x = torch.randn(10, 30, 512)
     wp = WaveletPacket2D(data=x,
                          wavelet=wavelet,
                          mode='reflect')
 
     wp_keys = list(product(["a", "h", "v", "d"], repeat=1))
     np_lst = []
     for key in wp_keys:
          np_lst.append(wp["".join(key)])
     viz = torch.cat([torch.cat([np_lst[0], np_lst[1]], -2),
                      torch.cat([np_lst[2], np_lst[3]], -2)], -1)
     print(viz.shape)
```
it prints ```torch.Size([10, 1, 30, 512])```

**Comment by snakers4** [comment]
Many thanks!

**Comment by v0lta** [comment]
This has now been explored in: https://arxiv.org/abs/2210.01989
